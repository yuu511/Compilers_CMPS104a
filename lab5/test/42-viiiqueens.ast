ROOT "<<ROOT>>" 0.0.0 {0} 
FUNCTION "" 6.22.1 {0} void function 
TYPE_ID "" 6.22.1 {0} 
VOID "void" 6.22.1 {0} 
IDENT "fail" 6.22.6 {0} 
PARAM "(" 6.22.11 {0} 
TYPE_ID "" 6.22.12 {1} string variable param lval 
STRING "string" 6.22.12 {0} 
IDENT "expr" 6.22.19 {0} 
TYPE_ID "" 6.22.25 {1} string variable param lval 
STRING "string" 6.22.25 {0} 
IDENT "file" 6.22.32 {0} 
TYPE_ID "" 6.22.38 {1} int variable param lval 
INT "int" 6.22.38 {0} 
IDENT "line" 6.22.42 {0} 
FUNCTION "" 6.24.1 {1} void function 
TYPE_ID "" 6.24.1 {0} 
VOID "void" 6.24.1 {0} 
IDENT "putchr" 6.24.6 {0} 
PARAM "(" 6.24.13 {0} 
TYPE_ID "" 6.24.14 {2} int variable param lval 
INT "int" 6.24.14 {0} 
IDENT "chr" 6.24.18 {0} 
FUNCTION "" 6.25.1 {2} void function 
TYPE_ID "" 6.25.1 {0} 
VOID "void" 6.25.1 {0} 
IDENT "putint" 6.25.6 {0} 
PARAM "(" 6.25.13 {0} 
TYPE_ID "" 6.25.14 {3} int variable param lval 
INT "int" 6.25.14 {0} 
IDENT "num" 6.25.18 {0} 
FUNCTION "" 6.26.1 {3} void function 
TYPE_ID "" 6.26.1 {0} 
VOID "void" 6.26.1 {0} 
IDENT "putstr" 6.26.6 {0} 
PARAM "(" 6.26.13 {0} 
TYPE_ID "" 6.26.14 {4} string variable param lval 
STRING "string" 6.26.14 {0} 
IDENT "str" 6.26.21 {0} 
FUNCTION "" 6.28.1 {4} int function 
TYPE_ID "" 6.28.1 {0} 
INT "int" 6.28.1 {0} 
IDENT "getchr" 6.28.5 {0} 
PARAM "(" 6.28.11 {0} 
FUNCTION "" 6.29.1 {5} string function 
TYPE_ID "" 6.29.1 {0} 
STRING "string" 6.29.1 {0} 
IDENT "getstr" 6.29.8 {0} 
PARAM "(" 6.29.14 {0} 
FUNCTION "" 6.30.1 {6} string function 
TYPE_ID "" 6.30.1 {0} 
STRING "string" 6.30.1 {0} 
IDENT "getln" 6.30.8 {0} 
PARAM "(" 6.30.13 {0} 
TYPE_ID "" 7.6.1 {0} int array variable lval vreg (7.6.1)
ARRAY "array" 7.6.1 {0} 
INT "int" 7.6.7 {0} 
IDENT "board" 7.6.12 {0} 
NULLPTR "nullptr" 7.6.20 {0} nullptr_t const 
FUNCTION "" 7.8.1 {7} int function 
TYPE_ID "" 7.8.1 {0} 
INT "int" 7.8.1 {0} 
IDENT "is_safe" 7.8.5 {0} 
PARAM "(" 7.8.13 {0} 
TYPE_ID "" 7.8.14 {8} int variable param lval 
INT "int" 7.8.14 {0} 
IDENT "newcol" 7.8.18 {0} 
BLOCK "{" 7.8.26 {8} 
TYPE_ID "" 7.9.4 {8} int variable local lval vreg (7.9.4)
INT "int" 7.9.4 {0} 
IDENT "col" 7.9.8 {0} 
INTCON "0" 7.9.14 {8} int const 
TYPE_ID "" 7.10.4 {8} int variable local lval vreg (7.10.4)
INT "int" 7.10.4 {0} 
IDENT "diagonal" 7.10.8 {0} 
INTCON "0" 7.10.19 {8} int const 
WHILE "while" 7.11.4 {8} 
LT "<" 7.11.15 {8} int vreg 
IDENT "col" 7.11.11 {8} int variable local lval (7.9.4)
IDENT "newcol" 7.11.17 {8} int variable param lval (7.8.14)
BLOCK "{" 7.11.25 {8} 
IF "if" 7.12.7 {8} 
EQ "==" 7.12.22 {8} int vreg 
INDEX "[" 7.12.16 {8} int lval vaddr 
IDENT "board" 7.12.11 {0} int array variable lval (7.6.1)
IDENT "col" 7.12.17 {8} int variable local lval (7.9.4)
INDEX "[" 7.12.30 {8} int lval vaddr 
IDENT "board" 7.12.25 {0} int array variable lval (7.6.1)
IDENT "newcol" 7.12.31 {8} int variable param lval (7.8.14)
RETURN "return" 7.12.40 {8} 
INTCON "0" 7.12.47 {8} int const 
'=' "=" 7.13.16 {8} int variable local lval vreg 
IDENT "diagonal" 7.13.7 {8} int variable local lval (7.10.4)
'-' "-" 7.13.29 {8} int vreg 
INDEX "[" 7.13.23 {8} int lval vaddr 
IDENT "board" 7.13.18 {0} int array variable lval (7.6.1)
IDENT "col" 7.13.24 {8} int variable local lval (7.9.4)
INDEX "[" 7.13.36 {8} int lval vaddr 
IDENT "board" 7.13.31 {0} int array variable lval (7.6.1)
IDENT "newcol" 7.13.37 {8} int variable param lval (7.8.14)
IF "if" 7.14.7 {8} 
EQ "==" 7.14.20 {8} int vreg 
IDENT "diagonal" 7.14.11 {8} int variable local lval (7.10.4)
'-' "-" 7.14.27 {8} int vreg 
IDENT "col" 7.14.23 {8} int variable local lval (7.9.4)
IDENT "newcol" 7.14.29 {8} int variable param lval (7.8.14)
RETURN "return" 7.14.37 {8} 
INTCON "0" 7.14.44 {8} int const 
IF "if" 7.15.7 {8} 
EQ "==" 7.15.20 {8} int vreg 
IDENT "diagonal" 7.15.11 {8} int variable local lval (7.10.4)
'-' "-" 7.15.30 {8} int vreg 
IDENT "newcol" 7.15.23 {8} int variable param lval (7.8.14)
IDENT "col" 7.15.32 {8} int variable local lval (7.9.4)
RETURN "return" 7.15.37 {8} 
INTCON "0" 7.15.44 {8} int const 
'=' "=" 7.16.11 {8} int variable local lval vreg 
IDENT "col" 7.16.7 {8} int variable local lval (7.9.4)
'+' "+" 7.16.17 {8} int vreg 
IDENT "col" 7.16.13 {8} int variable local lval (7.9.4)
INTCON "1" 7.16.19 {8} int const 
RETURN "return" 7.18.4 {8} 
INTCON "1" 7.18.11 {8} int const 
FUNCTION "" 7.21.1 {8} void function 
TYPE_ID "" 7.21.1 {0} 
VOID "void" 7.21.1 {0} 
IDENT "printqueens" 7.21.6 {0} 
PARAM "(" 7.21.17 {0} 
BLOCK "{" 7.21.20 {9} 
TYPE_ID "" 7.22.4 {9} int variable local lval vreg (7.22.4)
INT "int" 7.22.4 {0} 
IDENT "col" 7.22.8 {0} 
INTCON "0" 7.22.14 {9} int const 
WHILE "while" 7.23.4 {9} 
LT "<" 7.23.15 {9} int vreg 
IDENT "col" 7.23.11 {9} int variable local lval (7.22.4)
INTCON "8" 7.23.17 {9} int const 
BLOCK "{" 7.23.20 {9} 
CALL "(" 7.24.14 {1} void function (6.24.1)
IDENT "putchr" 7.24.7 {0} 
'+' "+" 7.24.26 {9} int vreg 
INDEX "[" 7.24.20 {9} int lval vaddr 
IDENT "board" 7.24.15 {0} int array variable lval (7.6.1)
IDENT "col" 7.24.21 {9} int variable local lval (7.22.4)
CHARCON "'1'" 7.24.28 {9} int const 
'=' "=" 7.25.11 {9} int variable local lval vreg 
IDENT "col" 7.25.7 {9} int variable local lval (7.22.4)
'+' "+" 7.25.17 {9} int vreg 
IDENT "col" 7.25.13 {9} int variable local lval (7.22.4)
INTCON "1" 7.25.19 {9} int const 
CALL "(" 7.27.11 {1} void function (6.24.1)
IDENT "putchr" 7.27.4 {0} 
CHARCON "'\n'" 7.27.12 {9} int const 
FUNCTION "" 7.30.1 {9} void function 
TYPE_ID "" 7.30.1 {0} 
VOID "void" 7.30.1 {0} 
IDENT "queens" 7.30.6 {0} 
PARAM "(" 7.30.13 {0} 
TYPE_ID "" 7.30.14 {10} int variable param lval 
INT "int" 7.30.14 {0} 
IDENT "newcol" 7.30.18 {0} 
BLOCK "{" 7.30.26 {10} 
TYPE_ID "" 7.31.4 {10} int variable local lval vreg (7.31.4)
INT "int" 7.31.4 {0} 
IDENT "row" 7.31.8 {0} 
INTCON "0" 7.31.14 {10} int const 
IF "if" 7.32.4 {10} 
EQ "==" 7.32.15 {10} int vreg 
IDENT "newcol" 7.32.8 {10} int variable param lval (7.30.14)
INTCON "8" 7.32.18 {10} int const 
CALL "(" 7.32.32 {8} void function (7.21.1)
IDENT "printqueens" 7.32.21 {0} 
BLOCK "{" 7.33.9 {10} 
WHILE "while" 7.34.7 {10} 
LT "<" 7.34.18 {10} int vreg 
IDENT "row" 7.34.14 {10} int variable local lval (7.31.4)
INTCON "8" 7.34.20 {10} int const 
BLOCK "{" 7.34.23 {10} 
'=' "=" 7.35.24 {10} int lval vreg vaddr 
INDEX "[" 7.35.15 {10} int lval vaddr 
IDENT "board" 7.35.10 {0} int array variable lval (7.6.1)
IDENT "newcol" 7.35.16 {10} int variable param lval (7.30.14)
IDENT "row" 7.35.26 {10} int variable local lval (7.31.4)
IF "if" 7.36.10 {10} 
CALL "(" 7.36.22 {7} int function (7.8.1)
IDENT "is_safe" 7.36.14 {0} 
IDENT "newcol" 7.36.23 {10} int variable param lval (7.30.14)
CALL "(" 7.36.39 {9} void function (7.30.1)
IDENT "queens" 7.36.32 {0} 
'+' "+" 7.36.47 {10} int vreg 
IDENT "newcol" 7.36.40 {10} int variable param lval (7.30.14)
INTCON "1" 7.36.49 {10} int const 
'=' "=" 7.37.14 {10} int variable local lval vreg 
IDENT "row" 7.37.10 {10} int variable local lval (7.31.4)
'+' "+" 7.37.20 {10} int vreg 
IDENT "row" 7.37.16 {10} int variable local lval (7.31.4)
INTCON "1" 7.37.22 {10} int const 
FUNCTION "" 7.42.1 {10} int function 
TYPE_ID "" 7.42.1 {0} 
INT "int" 7.42.1 {0} 
IDENT "main" 7.42.5 {0} 
PARAM "(" 7.42.9 {0} 
BLOCK "{" 7.42.12 {11} 
'=' "=" 7.43.10 {11} int array variable lval vreg 
IDENT "board" 7.43.4 {0} int array variable lval (7.6.1)
ALLOC "alloc" 7.43.12 {11} int array 
ARRAY "array" 7.43.18 {0} 
INT "int" 7.43.24 {0} 
INTCON "8" 7.43.31 {11} int const 
CALL "(" 7.44.11 {9} void function (7.30.1)
IDENT "queens" 7.44.4 {0} 
INTCON "0" 7.44.12 {11} int const 
RETURN "return" 7.45.4 {11} 
INTCON "0" 7.45.11 {11} int const 
