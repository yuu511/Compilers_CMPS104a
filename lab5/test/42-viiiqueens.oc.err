invalid condition for loop: 7.36.22
invalid condition for loop3ac: invalid statement passed :if
NUMBER OF ERRORS REPORTED: 2
