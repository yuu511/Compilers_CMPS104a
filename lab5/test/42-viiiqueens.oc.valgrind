==5920== Memcheck, a memory error detector
==5920== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==5920== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==5920== Command: ../oc 42-viiiqueens.oc
==5920== 
==5920== 
==5920== HEAP SUMMARY:
==5920==     in use at exit: 0 bytes in 0 blocks
==5920==   total heap usage: 1,141 allocs, 1,141 frees, 186,415 bytes allocated
==5920== 
==5920== All heap blocks were freed -- no leaks are possible
==5920== 
==5920== For counts of detected and suppressed errors, rerun with: -v
==5920== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
