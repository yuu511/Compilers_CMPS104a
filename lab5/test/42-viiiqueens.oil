board:     .global ptr nullptr
is_safe:   .function int 
           .param int newcol
           .local int col
           .local int diagonal
           col = 0  
           diagonal = 0  
.wh0:      goto .od0 if col > newcol
.do0:
.if0:      $t0:i = board[col * :i]  
           $t1:i = board[newcol * :i]  
           goto .fi0 if $t0:i != $t1:i
.th0:      return 0
.fi0:
           $t3:i = board[col * :i]  
           $t4:i = board[newcol * :i]  
           $t2:i = $t3:i - $t4:i
           diagonal = $t2:i  
.if1:      $t5:i = col - newcol
           goto .fi1 if diagonal != $t5:i
.th1:      return 0
.fi1:
.if2:      $t6:i = newcol - col
           goto .fi2 if diagonal != $t6:i
.th2:      return 0
.fi2:
           $t7:i = col + 1
           col = $t7:i  
           goto .wh0  
.od0
           return 1
           return
           .end
printqueens: .function 
           .local int col
           col = 0  
.wh1:      goto .od1 if col > 8
.do1:      $t1:i = board[col * :i]  
           $t0:i = $t1:i + '1'
           call putchr($t0:i)  
           $t2:i = col + 1
           col = $t2:i  
           goto .wh1  
.od1
           call putchr('\n')  
           return
           .end
queens:    .function 
           .param int newcol
           .local int row
           row = 0  
.if3:      goto .el3 if newcol != 8
.th3:      call printqueens()  
.fi3       goto   
.el3:
.wh2:      goto .od2 if row > 8
.do2:      board[newcol * :i] = row  
.if4:      $t0:i = call is_safe(newcol)  
           goto .fi4 if not $t0:i  
.th4:      $t1:i = newcol + 1
           call queens($t1:i)  
.fi4:
           $t2:i = row + 1
           row = $t2:i  
           goto .wh2  
.od2
.fi3:
           return
           .end
main:      .function int 
           $t0:p = 8 * sizeof int
           board = call malloc($t0:p)  
           call queens(0)  
           return 0
           return
           .end
