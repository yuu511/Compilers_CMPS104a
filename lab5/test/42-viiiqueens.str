string_set[   0]:    7296690267450279874 0x55daba0c9d58->"chr"
                   13348923631108115385 0x55daba0ca718->"putstr"
string_set[   5]:   12679825168387489523 0x55daba0ca278->"num"
string_set[   6]:   17899137375351314596 0x55daba0cea78->"8"
                    9317282556992397945 0x55daba0cc6d8->"if"
string_set[   7]:    6814028903552918164 0x55daba0cb498->"array"
                   10836626480094982265 0x55daba0cbe18->"col"
                    2297668033614959926 0x55daba0cbf58->"0"
string_set[  10]:    1583576014053361350 0x55daba0ca8d8->"str"
                   13497578857207247412 0x55daba0c9318->"string"
                   11597697714117577063 0x55daba0cb718->"="
string_set[  11]:   15403598725584096552 0x55daba0cc108->"diagonal"
                    1396094175042499165 0x55daba0cb558->"<"
string_set[  13]:   12868445110721718657 0x55daba0c9978->")"
string_set[  16]:   10159970873491820195 0x55daba0ce0e8->"1"
                   11911796782241914524 0x55daba0c8088->"void"
string_set[  18]:   13581035385482130682 0x55daba0cc898->"["
string_set[  19]:    5344662657899890615 0x55daba0c9a38->";"
string_set[  20]:   11868120863286193964 0x55daba0d10f8->"main"
                   15816506932648033945 0x55daba0c9258->"("
string_set[  25]:   11166701219533395393 0x55daba0caef8->"getstr"
string_set[  26]:   11526621952165666095 0x55daba0ce408->"printqueens"
string_set[  27]:    8374522807849457143 0x55daba0cfa18->"row"
                   12808232827189627082 0x55daba0cb7d8->"nullptr"
string_set[  29]:    1370730858159036685 0x55daba0ce028->"+"
                   14635384904416252381 0x55daba0ccfb8->"return"
string_set[  32]:    7110316202799739443 0x55daba0cb6d8->"board"
string_set[  34]:    8290819294854170237 0x55daba0d12b8->"alloc"
                    7211130878751977441 0x55daba0c9b98->"putchr"
string_set[  35]:   14736352883882276958 0x55daba0cccb8->"=="
string_set[  36]:    4132753638312158367 0x55daba0cbb18->"newcol"
string_set[  37]:    6192799629375248351 0x55daba0cf378->"'\n'"
                   12291392672316406109 0x55daba0c9898->"line"
string_set[  39]:   18413295556186398326 0x55daba0ca0b8->"putint"
string_set[  40]:   10319632053171904911 0x55daba0cc398->"while"
string_set[  43]:    7372855250431704371 0x55daba0cd2f8->"-"
string_set[  44]:      21606623943142546 0x55daba0cbcd8->"{"
string_set[  45]:    6142509188972423790 0x55daba0c9498->""
string_set[  47]:   13686721134187162920 0x55daba0ce238->"}"
string_set[  53]:   12051031859636400188 0x55daba0ccc58->"]"
string_set[  56]:    7574291458008592867 0x55daba0cf5d8->"queens"
string_set[  57]:    9386170080689500517 0x55daba0c93d8->"expr"
string_set[  58]:    4017852018407011111 0x55daba0c9658->"file"
                    3729804957429652673 0x55daba0c2d08->"<<ROOT>>"
string_set[  59]:   16579916199511856337 0x55daba0cb1f8->"getln"
string_set[  61]:   18004526203352653430 0x55daba0cb958->"is_safe"
string_set[  63]:    9611428142580219387 0x55daba0c9158->"fail"
string_set[  64]:   12597418120160561823 0x55daba0c97d8->"int"
string_set[  69]:    1873634190020599143 0x55daba0d0068->"else"
string_set[  70]:    9201800968376702500 0x55daba0cabf8->"getchr"
string_set[  71]:   14526079891305879776 0x55daba0cb698->">"
string_set[  77]:    8475192774821045282 0x55daba0cef78->"'1'"
string_set[  78]:    4706206694200543916 0x55daba0c9198->","
load_factor = 0.671
bucket_count = 79
max_bucket_size = 3
