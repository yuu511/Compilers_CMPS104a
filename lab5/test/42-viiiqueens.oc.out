CONDITIONAL WHILE STATEMENT SYMBOL TOK_LT 
CONDITIONAL IF STATEMENT SYMBOL TOK_EQ 
CONDITIONAL IF STATEMENT SYMBOL TOK_EQ 
CONDITIONAL IF STATEMENT SYMBOL TOK_EQ 
CONDITIONAL WHILE STATEMENT SYMBOL TOK_LT 
CONDITIONAL IF STATEMENT SYMBOL TOK_EQ 
CONDITIONAL WHILE STATEMENT SYMBOL TOK_LT 
CONDITIONAL IF STATEMENT SYMBOL TOK_CALL 
